rte/kernels-openacc/mo_rte_solver_kernels.F90

```diff
@@ -413,109 +414,138 @@ end subroutine sw_solver_noscat
-      call sw_two_stream(ncol, nlay, ngpt, mu0, &
-                         tau , ssa , g   ,      &
-                         Rdif, Tdif, Rdir, Tdir, Tnoscat)
-      call sw_source_2str(ncol, nlay, ngpt, top_at_1,       &
-                          Rdir, Tdir, Tnoscat, sfc_alb_dir, &
-                          source_up, source_dn, source_srf, flux_dir)
-      call adding(ncol, nlay, ngpt, top_at_1,   &
-                  sfc_alb_dif, Rdif, Tdif,      &
-                  source_dn, source_up, source_srf, flux_up, flux_dn)
-      !
-      ! adding computes only diffuse flux; flux_dn is total
-      !
-      !$acc  parallel loop collapse(3)
-      do igpt = 1, ngpt
-        do ilay = 1, nlay+1
-          do icol = 1, ncol
-            flux_dn(icol,ilay,igpt) = flux_dn(icol,ilay,igpt) + flux_dir(icol,ilay,igpt)
-          end do
+  subroutine sw_solver_2stream (ncol, nlay, ngpt, top_at_1, &
+                                tau, ssa, g, mu0,           &
+                                sfc_alb_dir, sfc_alb_dif,   &
+                                flux_up, flux_dn, flux_dir) bind (C, name="sw_solver_2stream")
+    integer,                               intent(in   ) :: ncol, nlay, ngpt ! Number of columns, layers, g-points
+    logical(wl),                           intent(in   ) :: top_at_1
+    real(wp), dimension(ncol,nlay,  ngpt), intent(in   ) :: tau, &  ! Optical thickness,
+                                                            ssa, &  ! single-scattering albedo,
+                                                            g       ! asymmetry parameter []
+    real(wp), dimension(ncol            ), intent(in   ) :: mu0     ! cosine of solar zenith angle
+    real(wp), dimension(ncol,       ngpt), intent(in   ) :: sfc_alb_dir, sfc_alb_dif
+                                                                  ! Spectral albedo of surface to direct and diffuse radiation
+    real(wp), dimension(ncol,nlay+1,ngpt), &
+                                            intent(  out) :: flux_up ! Fluxes [W/m2]
+    real(wp), dimension(ncol,nlay+1,ngpt), &                        ! Downward fluxes contain boundary conditions
+                                            intent(inout) :: flux_dn, flux_dir
+    ! -------------------------------------------
+    integer :: icol, ilay, igpt
+    real(wp), dimension(ncol,nlay,ngpt) :: Rdif, Tdif, Rdir, Tdir, Tnoscat
+    real(wp), dimension(ncol,nlay,ngpt) :: source_up, source_dn
+    real(wp), dimension(ncol     ,ngpt) :: source_srf
+    ! ------------------------------------
+    !
+    ! Cell properties: transmittance and reflectance for direct and diffuse radiation
+    !
+    !$acc enter data copyin(tau, ssa, g, mu0, sfc_alb_dir, sfc_alb_dif, flux_dn, flux_dir)
+    !$acc enter data create(Rdif, Tdif, Rdir, Tdir, Tnoscat, source_up, source_dn, source_srf, flux_up)
+    call sw_two_stream(ncol, nlay, ngpt, mu0, &
```

Comment:
These two routines - `sw_two_stream` and `sw_source_2_str` would also be good candidates for the loop fusion/stencil treatment, no? Is there a reason you did this for the LW sources and not here? Is the computation/memory profile different?

Reply:
They are indeed good candidates to be fused. The reason was that they have different loop pattern:  `sw_source_2_str` has layer dependency, and the inner loop has to be sequential. But it is still possible, and beneficial, to fuse these routines

Reply:
Let's come back to this.

Reply:
I've implemented the change, and it indeed helps. It is just a bigger refactoring, I'd rather open a new pull request for that

Reply:
OK, that's great, but indeed let's do it as part of a different PR.